### ISSUE TITLE
Add translations of language names to locale install list

### ISSUE THREAD

**Issue by amandastevens** [pkp/pkp-lib#7352]
**New problem description (3.4)**
The language lists (both in the software installer and e.g. on the language toggle plugin) list language names in the current language:

![image](https://user-images.githubusercontent.com/200411/186975142-bae0996b-eb88-4ec6-ba3c-a92e1312e9c2.png)

In cases where the same language is available in 2 character sets (e.g. Serbian), it's not clear which is which.

For users who find themselves using the software in a language they don't speak, they may not know their language's name in the current UI language.

Suggestion (Alec): Include the language name in both the current UI and listed language's own locale.

**Old problem description (3.3.x)**

When I install a new locale on an OJS 3.3 site as an Administrator in the Site Settings, the list of locales presented shows the locale names in their native script. If I can't read that language, it's difficult to identify which locale I need to install. I was only able to identify which locale was Simplified Chinese by consulting the registry/locales.xml file.

I suggest that the locale list show the language name in its native script and a translation of it into English, German, Spanish, Indonesian, etc., depending on the interface language being used.

This would be especially helpful for multi-OJS site administrators.

This is the current locale list: 
![Language list](https://user-images.githubusercontent.com/35044512/135270428-106bff7f-cc95-40b2-83fc-afa4a89e8df1.png)

**Comment by NateWr** [comment]
I'm not sure if we have every language translated into every other language. @asmecher do you know if this is part of the ISO library we're using?

(Also, I didn't know we have a Scottish Gaelic translation of OJS!! :joy:)

**Comment by asmecher** [comment]
The php-isocodes library we use for countries and currencies does include 639-3 language list. Note that it's harder to customize entries once they're delegated to that library, though; see https://github.com/asmecher/customCountryList for a plugin that allows adjustments to the country list.

**Comment by amandastevens** [comment]
Thanks, I didn't know about that plugin.

**Comment by asmecher** [comment]
In the `main` branch (pre-3.4), the language lists have now been converted to the current language (probably since https://github.com/pkp/pkp-lib/issues/6328). However, this isn't a great solution either. I've updated the issue to show both formulations of the problem and scheduled this issue for 3.4; for 3.3.x, you can customize `registry/locales.xml` as a work-around.

Some examples we can look to on major websites:

debian.org:
![image](https://user-images.githubusercontent.com/200411/186975991-3c206399-8c83-4e8f-bae3-d9767eb84f0d.png)

un.org:
![image](https://user-images.githubusercontent.com/200411/186976106-f4d6d216-ff9d-48db-bbee-fb8847b2f707.png)

wordpress.org (probably relying on https://github.com/pkp/pkp-lib/issues/7512):
![image](https://user-images.githubusercontent.com/200411/186976322-1b900aa3-7046-4fae-8107-6b29671a4a97.png)

Note that many major sites do not appear to have language selectors on them -- probably due to their reliance on [Accept-Language headers (7512)](https://github.com/pkp/pkp-lib/issues/7512), making it appear transparent to the user. However, while that would be nice for an automatic pre-selection, IMO there's value in making our multilingual support explicit.

**Comment by NateWr** [comment]
So is the idea to have the list show in the current locale, then it's own locale?

> **Change Language**
> ✔ English
> Arabic (عربى)
> Ukrainian (українська)

**Comment by asmecher** [comment]
Yes, IMO that would be the best result. The encoding would be implicit (e.g. Latin vs. Cyrillic) but I don't see any benefit in exposing that more explicitly.

**Comment by touhidurabir** [comment]
> So is the idea to have the list show in the current locale, then it's own locale?
>> Change Language
✔ English
Arabic (عربى)
Ukrainian (українська)

@asmecher  In that case do we plan to remove the `country` name ? Right now we shows as 
```
Arabic (Iraq)
Spanish (Spain)
Spanish (Mexico)
French (Canada)
French (France)
```

we have same language with different country . if we plan to remove the country, then it make no sense to have multiple language like `French` for `Canada` and `France` . in that case, we will have something like this 

```
Arabic (عربي)
Spanish (Español)
French (Français)
```

but I think that will be a breaking change . Should we do something like 

```
Arabic (Iraq)(العراق - عربي)
Spanish (Spain)(España - Español)
French (France)(France - Français)
```

so basically the format will be like 
`LANGUAGE_NAME_IN_ENGLISH(COUNTRY_NAME_IN_ENGLISH)(COUNTRY_NAME_IN_LANGUAGE_LOCALE - LANGUAGE_NAME_IN_LANGUAGE_LOCALE)`

what do you think ?

**Comment by NateWr** [comment]
We should remove the country name. This causes a fair amount of confusion and sometimes controversy.

**Comment by touhidurabir** [comment]
> We should remove the country name. This causes a fair amount of confusion and sometimes controversy.

@NateWr in that case how are we going to how the locale lists as `French (Canada)` as locale value `fr_CA` and `French (France)` has locale value `fr_FR` ? So if a journal/context has already installed `fr_CA` but in our list we can only show 
```
French (Français)
```
without country . so what value it should be ? we will have no way to show option for both value `fr_CA` and `fr_FR` .

**Comment by NateWr** [comment]
I did an informal survey on Mattermost with some large hosts and I found that this case is not really used. The only cases they found where two variants of the same language with the same character set were installed were a few `pt_BR` and `pt_PT`. The publishers thought this was likely an attempt at inclusion because before our language always included the country. 

In other words, they just wanted Portuguse but because our UI said "Portugese Brazil" and "Portugese Portugal" they felt they needed to enable both so that authors weren't led to believe they only supported Brazilian or Portugese authors.

That's what I mean about confusion. Countries are a bad proxy for language and rarely map onto real-world usage.

**Comment by asmecher** [comment]
@touhidurabir, the country is already excluded in the `main` branch; see https://github.com/pkp/pkp-lib/issues/7871 for details about that change.

**Comment by asmecher** [comment]
Touhidur and I talked about this today and especially the case where we do have to present several languages with different country codes alongside each other (e.g. pt_PT and pt_BR). This won't be common for end users but there is the case above that Nate mentioned (https://github.com/pkp/pkp-lib/issues/7352#issuecomment-1238251497) and it will be a problem e.g. on the configuration forms and installation form.

We decided that a good approach that covers all cases might be to examine the list before presentation to the user, determine whether there are multiple candidates for a language (e.g. with 2 countries), and if there are, then include the country name in the resulting list.

For most users, this will mean only showing the language name -- but for those with multiple candidates, it will also include the country.

**Comment by NateWr** [comment]
The case that I mentioned in my comment was actually a case _against_ using country names. Journals were enabling the same language in two countries because the country names made the language seem exclusive to one country or another.

Do we know of a real-world use case where a journal _wants_ to have two country options for the same language with the same character set?

**Comment by asmecher** [comment]
On the installation form (and the settings forms) we need something that'll distinguish between `fr_CA` and `fr_FR`, and can't just list "French" twice. So those lists need to be specific.

For end-user-facing content, they should be as general as possible -- listing only the language `Portuguese / Português` unless the journal has chosen to enable both `Portuguese / Português (Brazil)` and `Portuguese / Português (Portugal)`. In the latter case the country should be presented too, so it's not just `Portuguese / Português` twice. I agree that this is a boundary case but I think it would be nicer to go with an option that wouldn't leave existing users up a creek.

There probably aren't many cases where two countries should be listed for the same language, but I'm willing to bet one exists that'll be important to someone. Thinking of cases where...
-  the language is formally standardized in two places e.g. Swiss Standard German (de_CH): https://en.wikipedia.org/wiki/Swiss_Standard_German, and/or
- places where it's politically important to distinguish between locales (I'm sure you can think of a few of these).

@NateWr, for a more general solution to the issue that led a journal to enabling two Portugueses, I'd like to stop forcing language codes to specify countries in the first place, in line with standard locale code usage. See https://github.com/pkp/pkp-lib/issues/8210.

**Comment by NateWr** [comment]
Ok, makes sense. It won't "fix" things for existing users who have already enabled two languages to compensate. But it will improve things going forward. Maybe we can mention it in our release notebook, too.

**Comment by touhidurabir** [comment]
@asmecher please review the PRs 

[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/8289
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/3552

**Comment by asmecher** [comment]
@touhidurabir, this looks very good in the installation/settings forms when viewing the English case. When I switch over to French and view the list, it appears that many language names still appear in English (along with their own language name) -- do you see the same?
![image](https://user-images.githubusercontent.com/200411/192395930-e1ebb157-246c-414c-8c27-4fe15207b094.png)

Lastly, the language selector in the front end [needs to be adjusted too](https://github.com/pkp/pkp-lib/issues/7352#issuecomment-1248423344); at the moment, if you install both "Portuguese" variants, they'll be distinct and clear in the setup forms but then both translate to "Portuguese" in the front end. Example (in French)
![image](https://user-images.githubusercontent.com/200411/192396060-87a7d9ca-7e60-488a-a271-4a283feaf8df.png)

**Comment by touhidurabir** [comment]
> When I switch over to French and view the list, it appears that many language names still appear in English (along with their own language name)

@asmecher yes, this is happening for me also . I have checked in the `main` branch(which does not include this changes related to this issue PR yet), this is happening there also . For example , when I set `French(France)` as primary locale, can see the language `Armenian` when it should be `Arménien` for `French` . I guess may be some language name not getting translated properly. 

> the language selector in the front end https://github.com/pkp/pkp-lib/issues/7352#issuecomment-1248423344

it's done . 

Updated PRs

[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/8289
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/3552

**Comment by touhidurabir** [comment]
@asmecher updated PRs. Tests passed for all apps . 

[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/8289
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/3552
[omp](https://github.com/touhidurabir/omp) --> https://github.com/pkp/omp/pull/1214
[ops](https://github.com/touhidurabir/ops) --> https://github.com/pkp/ops/pull/364

**Comment by asmecher** [comment]
@touhidurabir, I still see a problem with the language toggle block plugin when both forms of Portuguese are enabled; they are not distinguished:
![image](https://user-images.githubusercontent.com/200411/193954946-856123c7-6fa0-4904-960c-9d17bf4495c6.png)

Also, I would recommend duplicating the `common.withForwardSlash` locale key for all `locale/*/common.po` files in pkp-lib. Normally I wouldn't suggest doing this but
- the translation that's used for English will work in 90% of cases, and
- when no translation is available, the language won't appear at all in the drop-down -- the user will only see `##common.withParenthesis##`, which will make it hard to navigate.

**Comment by touhidurabir** [comment]
@asmecher I have update all the locale presentation (if haven't missed any) with the new approach that includes both site and context wise display . Also updated the `locale/*/common.po` file . Please review . if all ok , then I will port the changes to other apps .

**Comment by asmecher** [comment]
@touhidurabir, the front-end display now shows too much information:
![image](https://user-images.githubusercontent.com/200411/194180611-7fbd7cb8-a90c-479d-8c23-6d236ce4a473.png)

See https://github.com/pkp/pkp-lib/issues/7352#issuecomment-1248423344 for a description of how the public-facing display should look.

For the above case, for example, it should be:
## Language
English
Français
Português (Brazil)
Português (Portugal)

Each language should be given in its own translation (thus "Français" rather than "French", even though the current UI language is English). The Brazil/Portugal distinction for Portuguese is shown because two Portuguese options are available; the region is not relevant for other languages (because there are not multiple forms of the same language enabled) so it should not be shown for them.

**Comment by touhidurabir** [comment]
@asmecher Please review the PRs . As per our last discussion, all requested changes have been implemented . 

[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/8289
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/3552

**Comment by jonasraoni** [comment]
Hiding the country when there's no need to disambiguation sounds good 👌

About enabling variants: I have to say the `pt_PT` sounds weird/wrong for me, a `pt_BR` speaker. I'd rather have an interface in `en_US` than `pt_PT`, so I think it makes sense to enable both :)

**Comment by asmecher** [comment]
@marcbria and @jonasraoni, this is what's already proposed/accepted in https://github.com/pkp/pkp-lib/issues/7352#issuecomment-1248423344.

**Comment by jonasraoni** [comment]
I've seen it, I'm just giving my support :)

**Comment by marcbria** [comment]
Sorry... I don't know if I read 7352 too fast or it's my poor reading comprehension in English, but I read you again and your comment was perfectly clear. 

Thank you Alec. I will report it back in the forum.

**Comment by asmecher** [comment]
Thanks, @touhidurabir! I've manually merged the two PRs into `main` (for OJS and pkp-lib) due to merge conflicts. However, I jumped the gun -- there are still a couple of things needing attention...
- There's a quirk with the language toggle block; it appears to always display the selected language's name in English. Do you see the same issue? Example:
  ![image](https://user-images.githubusercontent.com/200411/197906192-e5e083a1-1f13-4ebf-9f80-44a36f804586.png)
  In this screenshot the currently selected language is `Portuguese (Brasil)` -- note that it says `Portuguese` instead of `Português`. Switching over to `Português (Portugal)` causes the error to appear there (as with English and French).
- Could you port the OJS repo changes to OMP and OPS?

**Comment by asmecher** [comment]
On second thought, I'll take care of the port to OPS and OMP, but the English-language presentation is yours to look at!

**Comment by touhidurabir** [comment]
> English-language presentation is yours to look at!

@asmecher looking into it .

**Comment by marcbria** [comment]
That's great!

@asmecher, to properly report back in forum: The change will be for 3.4 or greater, isn't it?
If yes, any plans to port back to 3.3?

Thanks!

**Comment by asmecher** [comment]
This will be in 3.4, but won't be back-ported to 3.3.

**Comment by NateWr** [comment]
It looks like work on this has been merged to `main`. However, I just looked at an up-to-date checkout of `main` and the country names are showing up in the form language switcher even though I only have one country of a language enabled.

![language-country](https://user-images.githubusercontent.com/2306629/198249928-9eead9a7-28ef-4893-9153-600f6f4ddb71.png)

In the screenshot above, my active locale is `en_US`. I would expect to see:

![language](https://user-images.githubusercontent.com/2306629/198250314-bd9e7696-30de-4533-98bc-f11667e6e538.png)

Sorry if this is known and still being worked on.

**Comment by touhidurabir** [comment]
@NateWr good catch, it seems I missed that part to update . Also there is still one small change that I have missed and working on it . But do you think we should show it in primary locale or based on each language locale ? for example , for the user top right menu , we are showing each language name based on that language locale , check the screenshot of my local installation 
<img width="273" alt="CleanShot 2022-10-27 at 15 57 12@2x" src="https://user-images.githubusercontent.com/2587979/198254621-f77c7cd1-30a3-4e40-ba00-84d9cc648c0e.png">

I think the country name should be removed unless there is 2 same language (as in that case, we need some way to differentiate it) , So my guess it it should be like 
```
Français | English
```
or 
```
French/Français | English/English
```
Please correct me if I misunderstood it .

**Comment by NateWr** [comment]
In the forms we should always show the language name in the current locale. So in English it would be:

> French | English

This is different from the other places, because here the user is entering information rather than switching languages, and we assume only that they know the language they are viewing the UI in.

**Comment by NateWr** [comment]
Another place to check is in the form field labels. In a multilingual form, the fields in languages other than the current language have a special label:

```
msgid "form.multilingualLabel"
msgstr "{$label} in {$localeName}"
```

So, when viewing the form in English, the label for the French input field for "Title" will be:

> Title in French

Here, too, the language name should be shown in the current language.

**Comment by touhidurabir** [comment]
@asmecher please review the following PRs which add patch for missing changes . 

[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/8395
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/3599

**Comment by asmecher** [comment]
Merged, thanks!

**Comment by asmecher** [comment]
@touhidurabir, is this ready to be closed from your perspective?

**Comment by touhidurabir** [comment]
> is this ready to be closed from your perspective?

@asmecher I believe so . we can re-open again if more changes require again in some other place .

**Comment by asmecher** [comment]
(@NateWr, please let us know if you need more on this issue to resolve the presentation as it shows in the new submission wizard.)

**Comment by touhidurabir** [comment]
Based on feedback from @NateWr , some more fixing/patching needed for this one as there persist problem in the following cases 

1. In the publication forms (Submissions > Submission > Publication > Title & Abstract / Metadata / Permissions & Disclosure / Issue).
2. In the contributor forms (same location, but Contributors)
3. In the submission wizard (Submissions > New Submission)
4. In the file uploads form (Submissions > Submission > Upload File > Review Details)
5. In the issue forms (Issues > click issue > Issue Data
6. In the distribution settings (Settings > Distribution > License / Search Indexing )
7. In the submission settings (Settings > Workflow > Submission > Author Guidance)
8. In the file component settings (Settings > Workflow > Submission > Components > Add a Component)
9. In the review settings (Settings > Workflow > Review > Reviewer Guidance and Review Forms > Create Review Form)
10. In the publisher library (Settings > Workflow > Publisher Library > Add File)
11. When adding/editing an email template (Settings > Workflow > Emails > Add and edit templates > Click Edit for Discussion (Copyediting) > Add template)
12. Add navigation menu item (Settings > Website > Navigation > Add Item)
13. Add announcement and announcement types (First enable announcements at Settings > Website > Announcements to see the main navigation item on the left. Then go to Announcements > Add Announcement / Add Announcement Type.)
14. Create and edit section (Settings > Journal > Sections > Create Section / Categories > Create Categories)

need to update the generalised solution to address one overall implementation for all above cases and if possible, without the changing the view based implementation .

**Comment by asmecher** [comment]
One more: in the administration area, when you use the Install Locale button to select a new language to install, the raw locale codes should be shown (along with the names) to distinguish between ambiguities.

![image](https://user-images.githubusercontent.com/200411/215229476-672b21fa-0d1c-424e-aec7-dae0afa0e938.png)

Here's an example of an ambiguous entry -- they are for `uz_UZ@cyrillic` and `uz_UZ@latin`.

![image](https://user-images.githubusercontent.com/200411/215229612-9ba9b3a6-bd40-4e94-a7b9-284327b8b387.png)

**Comment by touhidurabir** [comment]
@asmecher  Patch PR . If the tests get passed and all ok, I will forward port it to `OMP/OPS`. I personally think it now has better implementation than before . 

[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/8589
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/3736

**Comment by asmecher** [comment]
We still have [this problem](https://github.com/pkp/pkp-lib/issues/7352#issuecomment-1407216035) on the installation form; can you add the locale codes there as well?

**Comment by touhidurabir** [comment]
> the raw locale codes should be shown (along with the names) to distinguish between ambiguities.
> We still have https://github.com/pkp/pkp-lib/issues/7352#issuecomment-1407216035 on the installation form; can you add the locale codes there as well?

 @asmecher already added . 

<img width="1270" alt="CleanShot 2023-02-02 at 10 27 48@2x" src="https://user-images.githubusercontent.com/2587979/216231883-af202ad1-2df9-4aca-acd3-f44b85824c5a.png">

**Comment by touhidurabir** [comment]
@asmecher updated PRs

[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/8589
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/3736

unless we want to make any more changes, it should be good to ported to `OMP/OPS` .

**Comment by touhidurabir** [comment]
@asmecher PRs

[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/8589
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/3736
[ops](https://github.com/pkp/ops) --> https://github.com/pkp/ops/pull/462
[omp](https://github.com/pkp/omp) --> https://github.com/pkp/omp/pull/1327

**Comment by asmecher** [comment]
All merged -- thanks, @touhidurabir!

**Comment by asmecher** [comment]
@touhidurabir, I still see the install form showing with ambiguous names, e.g. Uzbek:
![image](https://user-images.githubusercontent.com/200411/222868153-05086a80-9132-4d37-a57e-f856470c0798.png)

**Comment by asmecher** [comment]
Also, I think the original goal of this (described at the top) has been lost -- the language selection drop-down in the back-end is no longer showing translated language names. 
![image](https://user-images.githubusercontent.com/200411/222874437-ae5cf573-8109-4dbf-8364-ed728a61b345.png)

**Comment by touhidurabir** [comment]
> I still see the install form showing with ambiguous names, e.g. Uzbek:
> ![image](https://user-images.githubusercontent.com/200411/222868153-05086a80-9132-4d37-a57e-f856470c0798.png)

@asmecher this is how it's looks for me 

<img width="393" alt="CleanShot 2023-03-06 at 17 19 32@2x" src="https://user-images.githubusercontent.com/2587979/223096681-597fd4ce-0a52-43e0-bd01-4d020961bf52.png">

here `Uzbek` is shown as `Uzbek/Uzbek` as the translated name of `Uzbek` returns as `Uzbek` by our used package `sokil/php-isocodes` .  even setting the current lang to `uz` does not impact , run the following code 

```php
putenv('LANGUAGE=uz_UZ.UTF-8');
putenv('LC_ALL=uz_UZ.UTF-8');
setlocale(LC_ALL, 'uz_UZ.UTF-8');

use \Sokil\IsoCodes\IsoCodesFactory;

$isoCodes = new IsoCodesFactory();

$languages = $isoCodes->getLanguages();

$language = $languages->getByAlpha2('uz');

dump($language->getLocalName()); // print Uzbek

// $isoCodes->getCountries()->getByAlpha2('UZ')->getLocalName();
```
 Also as we have removed the country postfix e.g. `__2_LETTER_COUNTRY_CODE`, we have no way to determine the country code from lang code only so no country name also shown . 



> I think the original goal of this (described at the top) has been lost -- the language selection drop-down in the back-end is no longer showing translated language names.
> ![image](https://user-images.githubusercontent.com/200411/222874437-ae5cf573-8109-4dbf-8364-ed728a61b345.png)

Yes, thats got reverted back in our last patch PR . But as we have some temp cache via `static` variable for the `Site` class, we have to fallback to direct use of `Locale::getFormattedDisplayNames` in one place .

**Comment by touhidurabir** [comment]
@asmecher PRs to review 

[pkp-lib](https://github.com/pkp/pkp-lib) --> https://github.com/pkp/pkp-lib/pull/8742
[ojs](https://github.com/pkp/ojs) --> https://github.com/pkp/ojs/pull/3812

I will forward port to `OMP/OPS` once we finalize .

**Comment by asmecher** [comment]
@touhidurabir, are you looking at e.g. the main OJS installation form? The top of the page looks like this:
![image](https://user-images.githubusercontent.com/200411/223592718-9d59d5f6-2c34-4e1c-9ec1-3b96129674e3.png)

I'm still not seeing the locale codes there, which should be added to help disambiguate between similarly-named locales.

**Comment by touhidurabir** [comment]
> referring to the locale list with checkboxes, not the dropdown in the screenshot above.

added . But I think we decided not to make any modification in the installation page 

<img width="271" alt="CleanShot 2023-03-08 at 14 00 03@2x" src="https://user-images.githubusercontent.com/2587979/223655330-0d5525fc-5566-4ed7-8074-94f47c8a2a31.png">

**Comment by asmecher** [comment]
@touhidurabir, thanks, that looks good; I've merged to `main`, and the OMP and OPS parts were trivial enough so I patched those apps too.
